whylog/constraints/__init__.py

```diff
@@ -151,7 +151,27 @@ def _check_optional_params(self, correct_param_names, actual_param_names):
             raise ConstructorParamsError(self.TYPE, correct_param_names, actual_param_names)
 
     def verify(self, group_contents, param_dict):
-        # TODO remove mock
+        """
+        :param group_contents
+        group_contents[0] = date supposed to be earlier
+        group_contents[1] = date supposed to be later
+        :param param_dict
+        contains 'min_delta' and 'max_delta', time interval between
+        two given dates should satisfy requirement:
+        min_delta <= time delta <= max_delta
+        """
+        earlier_date, later_date = group_contents
+        actual_delta = later_date - earlier_date
+        param_min_delta = param_dict.get(self.MIN_DELTA)
```

Comment:
this is one of the most performance critical sections of the entire project

I suggest that the function (actually one of 3 functions: verify_min, verify_max and verify_both) is added to the object in the constructor, so that the `if param_xxx is not None`, the `param_dict.get` and `datetime.timedelta` are called once in the lifespan of constraint object, not millions of times. The function would be added to the object by using `self.verify = function`.